### ISSUE TITLE
callback(success, res or err) => callback(err, res or undefined)

### ISSUE THREAD

**Issue by vvo** [algolia/algoliasearch-client-js#42]
## tl;dr;
```js
var client = algoliasearch({
  applicationID: '',
  apiKey: ''
});

var index = client.initIndex('indexName');

index.search('YAW callback', function searchDone(err, res) {
  if (err) {
    // do something with err, it's an instance of `Error`
    return;
  }

  // everything is cool, use response
  console.log(res.hits);
})
```

## Boring stuff
The current callback convention for our JavaScript client is `index.search(callback)` where callback will be called with `callback(success, res or err)`.

A popular way to handle asynchronous JavaScript is to use the `callback(err or null, data or undefined)` convention.

Notable projects using or leveraging this callback convention:
- the whole [Node.js api](http://nodejs.org/api/)
- [caolan/async](https://github.com/caolan/async)
- [request/request](https://github.com/request/request#requestoptions-callback)
- most of the npm modules doing asyncrhonous stuff will use this convention

In the days of ["isomorphic JavaScript"](http://isomorphic.net/), we must stick with a well used convention to not fall into the "guys using fancy callback pattern" basket.

## TODO
- [x] move callback convention to `callback(err, res)`
- [x] note breaking changes in the issue if any

## References
- http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony
- http://caolanmcmahon.com/posts/nodejs_style_and_structure/
- http://nodeguide.com/style.html#callbacks

Linked with algolia/algoliasearch-client-node#18.
